### ISSUE TITLE
win32 buildbot status

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7458]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| moonfallen reported|
|-|-|
|Trac ID|trac#142|
|Type|enhancement|
|Created|2003-07-28 23:56:29Z|

Attachments:

* [tracebacks.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a2/2a2b47bf21a372f267deccbb420567f3d450b3c0/5621976dd831c05e6417c8702aa898d2371653b9.txt) (6539 bytes) - added by moonfallen on 2003-07-28 23:56:29Z - 
* [tb2.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a2/2a2b47bf21a372f267deccbb420567f3d450b3c0/31b11fb5dcef5228cfb10e313db8bc3e5d2afdca.txt) (1747 bytes) - added by moonfallen on 2003-07-29 06:00:18Z - 
* [slavecommand_diff.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a2/2a2b47bf21a372f267deccbb420567f3d450b3c0/bec3ed39e1ca5230e7ae811721679d2d13b1ef93.txt) (857 bytes) - added by norfolk on 2004-03-13 10:20:57Z - 
* [win32_buildbot_fixes.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a2/2a2b47bf21a372f267deccbb420567f3d450b3c0/65aa8db4b4572c36712ae3453bd57efbbb142cb3.txt) (1396 bytes) - added by moonfallen on 2004-03-13 11:46:00Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__142 142
type__enhancement enhancement
reporter__moonfallen moonfallen
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1059436589000000 1059436589000000
changetime__1100970317000000 1100970317000000
version__ 
owner__ 
cc__glyph cc__moonfallen cc__spiv cc__itamarst cc__warner cc__norfolk
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
(Note: I selected "wish" priority even though this doesn't
really fit.  Something like "task" would be more appropriate.)

Starting with buildbot 0.3.3, I created bot-w32 in my home
directory (c:\cygwin\home\cory\bot-w32) and created a tap using:

$ python c:/python22/scripts/mktap buildbot slave --name
bot-w32 -p xxxx -d bot-w32 -m pyramid.twistedmatrix.com:9987 

Then I ran it using:

$ python c:/python22/scripts/twistd -f buildslave-bot-w32.tap

This results in the traceback labeled (1) in the attached
file.  win32 has processes, so I don't know why this error
occurs.  I switched to the win32 reactor:

$ python c:/python22/scripts/twistd -r win32 -f
buildslave-bot-w32.tap

This results in the traceback labelled (2) in the attached
file, almost immediately.  I suspect this may be an
incompatibility between my version of buildbot and pyramid's
version of buildbot; that is, I need to load CVS.  At this
time, I cannot, because sourceforge sucks.  (I will retry
later).

Then I initiate a build by clicking on the 'win32' link at
http://twistedmatrix.com/buildbot

This results in two exceptions during the CVS checkout, one
right after another.  They are labelled (3) and (4) in the
attached file.

At this point buildbot hangs and won't do any more work.

Any revelations about any of the 4 tracebacks I'm seeing
would be appreciated.  When sourceforge comes back I'll try
CVS buildbot.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
[1] is the "if you want this, talk to itamar" catch in the
default reactor's spawnProcess when run on w32

[2] is a harmless buildmaster/buildslave version skew issue

[3] is a w32 reactor problem that itamar says is now fixed

[4] is a buildbot-being-nonportable issue that I say I've
now fixed

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Current status: running.  I will leave it up over night,
knock on wood it's still here tomorrow morning.  

Tracebacks 2, 3 and 4 have gone by-by thanks to checkins by
itamar to Twisted and warner to buildbot.  I did see the
traceback labeled (5) (attached) at one point, probably a
network burp or something.

Things that had to be changed:

- On my particular win32 configuration, there is no
python2.2; only python.
- build_ext must be done with -c mingw32 when msvc is not
installed and Cygwin is.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
New traceback with the win32eventreactor reactor.

2003/07/28 17:20 Pacific Daylight Time
[*buildslave-bot-w32*]   command 'pythonbin/trial -to -r
win32eventreactor twisted.test' in dir C:\cygwin\home\cory\bot-
w32\W32-full2.2\Twisted [24 681620]
2003/07/28 17:43 Pacific Daylight Time
[*buildslave-bot-w32*] Traceback (most recent call last):
          File
"C:\python22\Lib\site-packages\twisted\python\context.py",
line 53, in callWithContext
            return
self.currentContext().callWithContext(ctx, func, *args, **kw)
          File
"C:\python22\Lib\site-packages\twisted\python\context.py",
line 32, in callWithContext
            return func(*args,**kw)
          File
"c:\python22\Lib\site-packages\twisted\internet\default.py",
line 122, in run
            self.mainLoop()
          File
"c:\python22\Lib\site-packages\twisted\internet\default.py",
line 130, in mainLoop
            self.runUntilCurrent()
        --- &lt;exception caught here> ---
          File
"C:\python22\lib\site-packages\twisted\internet\base.py",
line 414, in runUntilCurrent
            call.func(*call.args, **call.kw)
          File
"c:\python22\Lib\site-packages\buildbot\slavecommand.py",
line 212, in doTimeout
            msg += ", killing pid %d" % self.process.pid
        exceptions.AttributeError: 'NoneType' object has no
attribute 'pid'

This is apparently in response to a timeout, as it doesn't
look like win32eventreactor ever finishes.

BTW, I'm already running with -r win32, isn't that the same
reactor?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
running 'twistd -r win32' just says which reactor the
buildslave itself will use. The tests are run in a separate
process, and trial is given various --reactor options to use
various reactors for the tests.

The traceback in msg328 is from a timeout confusion,
probably killing off a child that's already been removed.
I'll look into it further.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
For the last 8 hours or so the buildslave has been
connecting and disconnecting repeatedly, once per hour. 
This looks a lot like the issue itamar originally hit.  What
background do we have about this?

Also, I got the msg328 traceback again early this morning.

I have shut down the buildslave temporarily; I'll bring it
back up when someone is available to help debug this
disconnecting problem.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Looks like after exactly one hour of idle time, it
disconnects and reconnects.  Here's a sample of the log.

2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> disconnected.
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> will retry in 2 seconds
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] lost remote
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] Stopping factory
&lt;twisted.spread.pb.PBClientFactory instance at 0x00AA9928>
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] starting connect
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] Starting factory
&lt;twisted.spread.pb.PBClientFactory instance at 0x00AAE9C0>
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] message from master: attached
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> successfully connected.
2003/07/29 13:15 Pacific Daylight Time
[*buildslave-bot-w32*] builder 'win32' message from master:
attached
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> disconnected.
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> will retry in 2 seconds
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] lost remote
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] Stopping factory
&lt;twisted.spread.pb.PBClientFactory instance at 0x00AAE9C0>
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] starting connect
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] Starting factory
&lt;twisted.spread.pb.PBClientFactory instance at 0x01084710>
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] message from master: attached
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] &lt;buildbot.bot.Bot instance at
0x00A4A928> successfully connected.
2003/07/29 14:15 Pacific Daylight Time
[*buildslave-bot-w32*] builder 'win32' message from master:
attached

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
oh, right. 1 hour is the default keepalive interval if
TCP-level keepalives aren't available (i.e. under windows).

Try creating the buildslave .tap file with --keepalive=600 .
That will send a keepalive every 10 minutes. The goal is to
figure out what the NAT box that sits between you and the
rest of the world uses as a timeout value and set the
keepalive timer to be slightly smaller.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
With 10 minute keepalives, it hasn't timed out in the last
90 minutes (which it spent idle).  I will continue to keep
an eye on it, knock on wood.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Buildbot has performed well.  I have a couple of wishlist
items for a completely separate pair of buildslaves:

- python2.2 win32 builder
- python2.3 win32 builder

Also, buildbot is shut down.  I don't own the machine it was
running on, so I can't dedicate it to Twisted.  Glyph has
said he has a box for this; whenever he gets me VNC access
to it I'll be happy to set it up the fiddly details of
making buildslaves run.

I'd prefer win2k on that box, if there's a choice.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Still waiting on glyph to continue this.  Any progress with
the firewall details?

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom commented|
|-|-|

```
#!html
<pre>
Pretty please let moonfallen do it. I can't run tests very well with iocpreactor
if some of them fail on Windows regardless of reactor.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Used tap2ntsvc to build an ntsvc for buildbot-to-tm.com and sent it to norfolk.
 It's in his hands now. :-)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Buildbot is currently installed and running on norfolk's machine.  I forgot to
include "install CVS" in my instructions, though.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
warner, hate to bug you on this again but we're encountering an issue that's new
since the last time I set up a buildbot.  We connected to the tm.com buildbot
with a slave running buildbot code from CVS, yesterday's.  It checks out CVS
successfully but then hangs after putting "program finished with exit code 0" in
the log.  The log file is not closed yet, so I think the server is waiting for
the client to do something before sending the next command.

Any idea what could cause this?  Or how we could approach debugging it?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Warner, I think this is a bug in buildbot.  When I ran twistd -b I got this:


Jumping into debugger for post-mortem of exception '[Errno 3] The system cannot
find the path specified: 'source/*.*'':

slavecommand.py:
475                 # now copy tree to workdir
476                 if runtime.platformType != "posix":
477  ->                 shutil.copytree(self.cvsdir, self.workdir)
478                     return self.copyComplete(None)

self.cvsdir == 'source', but os.getcwd() is not the slave dir.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="norfolk's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|norfolk commented|
|-|-|

```
#!html
<pre>
I have a fix that allows it to get past the copytree problem, and move on to
other problems. :)   My change is in slavecommand_diff.txt, which I hope will be
attached to this issue when I submit this.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Applied norfolk's patch.  Ran again, CVS checks out successfully.  Next step
fails here:

"""

Jumping into debugger for post-mortem of exception '(3, 'CreateProcess', 'The
system cannot find the path specified.')':
> c:\python23\lib\site-packages\twisted\internet\win32eventreactor.py(296)__init__()
-> self.hProcess, hThread, dwPid, dwTid = win32process.CreateProcess(None,
cmdline, None, None, 1, 0, environment, path, StartupInfo)
(Pdb) list
291             win32file.CloseHandle(self.hStdinW)
292             self.hStdinW = tmp
293     
294             # create the process
295             cmdline = "%s %s" % (command, string.join(args[1:], ' '))
296  ->         self.hProcess, hThread, dwPid, dwTid =
win32process.CreateProcess(None, cmdline, None, None, 1, 0, environment, path,
StartupInfo)
297     
298             # close handles which only the child will use
299             win32file.CloseHandle(hStderrW)
300             win32file.CloseHandle(hStdoutW)
301             win32file.CloseHandle(hStdinR)
(Pdb) cmdline
'/bin/sh -c python  setup.py build_ext -c mingw32 -i'
 ^^^^^^^
"""

Whoops.


Patch attached.. runs %COMSPEC% /c instead of /bin/sh -c.  This works for the
specific case of the Twisted tests, but cmd is not sh, and users of buildbot
forget that at their peril. :)

The new patch combines with (obsoletes) norfolk's last patch.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
norfolk, you should also install pyOpenSSL 0.5.1 (available from
twisted.sf.net/contrib/).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Oh, and this is for both of you. It'd be nice if it also did a test run of the
win32eventreactor and (maybe even the IOCP reactor) in addition to default reactor.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Oops, look like it does win32 reactor. A comment for warner - the compile is
marked has having errors apparently because of a failing conftest.c? Those
should probably be ignored since they are used to test for features.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
Found another way for it to hang:

2004/03/13 11:29 Pacific Standard Time [Broker,client]   command '['cvs', '-z3',
'update', '-dP']' in dir C:\cygwin\home\PASS
Program\cvs\buildbot\W32-full2.2\source [None]
2004/03/13 11:29 Pacific Standard Time [-] command finished with signal None,
exit code 0
Jumping into debugger for post-mortem of exception '[Errno 17] File exists:
'C:\\cygwin\\home\\PASS Program\\cvs\\buildbot\\W32-full2.2\\Twisted'':

This probably only happened because I hit ^C in the middle of a buildbot run,
but it's not very graceful nonetheless.  Buildbot should clean out this
directory if it has to.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
i've committed the command-execution win32-compat patch to buildbot.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I think any further general discussion like this should be on the buildbot
masters list? anyway, resolving.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
